### ISSUE TITLE
synchronously check if item is cached

### ISSUE THREAD

**Issue by heyalexchoi** [Haneke/HanekeSwift#311]
is there no way to synchronously check if an item is cached? i am particularly interested in this operation for the disk cache.

if not, i can add it. is there any reason this feature would not be desirable?

essentially i would want to provide an accessor on Cache to check if a file exists at a path for a key.

**Comment by dcharbonnier** [comment]
Any asynchronous call can be synchronized on your side and generally speaking accessing storage synchronously does not seams like a good approach. What is preventing you from synchronization on your side and what is the use casse ?

**Comment by heyalexchoi** [comment]
the use case is to check if a resource exists in the cache before making a network request to get it.

as far as i can tell, and please correct me if i'm wrong, there is no way to see if an item is cached without actually fetching for it. if this is true, it seems unnecessarily expensive to load the whole object when i really just want to see if it exists.

i was planning on adding some methods to get the path for a key, then check with the file manager to see if a file exists at the path. i assumed this would be quick enough to do synchronously which would also simplify the interaction, but if others think it should be asynchronous, i could do that too.

**Comment by coolnalu** [comment]
@dcharbonnier a simple and quick way to check an image is loaded would be useful, so the image can set right away on the UI thread. Right now it is pessimistic so we put a placeholder in place but it quickly got replaced by the real image. This causes flickering because there is a short delay and the UI looks choppy from those.

**Comment by ThoseGuysInTown** [comment]
yes this would be a great feature! Did ya ever end up writing it?

**Comment by drewolbrich** [comment]
This seems to work. I am using it to avoid the flickering issue that @coolnalu describes above.

```
import Haneke

extension Haneke.Shared {
    
    /// Returns `true` if the image with the specified URL already exists in the image cache.
    static func imageCacheContainsImage(withURL url: URL) -> Bool {
        guard let (_, _, diskCache) = Haneke.Shared.imageCache.formats[HanekeGlobals.Cache.OriginalFormatName] else {
            return false
        }
        let fetcher = Haneke.NetworkFetcher<UIImage>(URL: url)
        let path = diskCache.path(forKey: fetcher.key)
        return FileManager.default.fileExists(atPath: path)
    }
    
}
```
